include/zephyr/drivers/mspi.h

```diff
@@ -0,0 +1,791 @@
+ * The controller can identify its devices and determine whether the access is
+ * allowed in a multiple device scheme.
+ */
+struct mspi_dev_id {
+	/** @brief device gpio ce */
+	struct gpio_dt_spec     ce;
+	/** @brief device index on DT */
+	uint16_t                dev_idx;
+};
+
+/**
+ * @brief MSPI controller configuration
+ */
+struct mspi_cfg {
+	/** @brief mspi channel number */
+	uint8_t                 ui8MSPIChannel;
+	/** @brief Configure operaton mode */
+	enum mspi_op_mode       eOPMode;
+	/** @brief Configure duplex mode */
+	enum mspi_duplex        eDuplex;
+	/** @brief DQS support flag */
+	bool                    bDQS;
+	/** @brief Software managed multi-slave enable */
+	bool                    bSwMultiSlave;
+	/** @brief GPIO chip-select line (optional) */
+	struct gpio_dt_spec     *pCE;
+	/** @brief GPIO chip-select line numbers (optional) */
+	uint32_t                ui32CeGpioNum;
+	/** @brief Slave number from 0 to host controller slave limit. */
+	uint32_t                ui32SlaveNum;
+	/** @brief Maximum supported frequency in MHz */
+	uint32_t                ui32MaxFreq;
+	/** @brief Whether to re-initialize controller */
+	bool                    bReinit;
+};
+
+/**
+ * @brief MSPI DT information
+ */
+struct mspi_dt_spec {
+	/** @brief MSPI bus */
+	const struct device     *bus;
+	/** @brief MSPI hardware specific configuration */
+	struct mspi_cfg         config;
+};
+
+/**
+ * @brief MSPI controller device specific configuration
+ */
+struct mspi_dev_cfg {
+	/** @brief Configure CE0 or CE1 */
+	uint32_t                ui32CENum;
+	/** @brief Configure frequency */
+	uint32_t                ui32Freq;
+	/** @brief Configure I/O mode */
+	enum mspi_io_mode       eIOMode;
+	/** @brief Configure data rate */
+	enum mspi_data_rate     eDataRate;
+	/** @brief Configure clock polarity and phase*/
+	enum mspi_cpp_mode      eCPP;
+	/** @brief Configure transfer endian */
+	enum mspi_endian        eEndian;
+	/** @brief Configure chip enable polarity */
+	enum mspi_ce_polarity   eCEPolarity;
+	/** @brief Configure DQS mode */
+	bool                    bDQSEnable;
+	/** @brief Configure number of clock cycles between
+	 * addr and data in RX direction
+	 */
+	uint32_t                ui32RXDummy;
+	/** @brief Configure number of clock cycles between
+	 * addr and data in TX direction
+	 */
+	uint32_t                ui32TXDummy;
+	/** @brief Configure read instruction */
+	uint32_t                ui32ReadInstr;
+	/** @brief Configure write instruction */
+	uint32_t                ui32WriteInstr;
+	/** @brief Configure instruction length */
+	uint16_t                ui16InstrLength;
+	/** @brief Configure address length */
+	uint16_t                ui16AddrLength;
+	/** @brief Configure memory boundary */
+	uint32_t                ui32MemBoundary;
+	/** @brief Configure break time */
+	uint32_t                ui32BreakTimeLimit;
+};
+
+/**
+ * @brief MSPI controller XIP configuration
+ */
+struct mspi_xip_cfg {
```

Comment:
General question here- how do you envision supporting memory mapped access to MSPI peripherals? This is often the most efficient (and programmatically simple) method to configure XIP or interfacing with something like PSRAM.

Reply:
I will leave the rest for Swoft as he is not on my time zone but he has written a full Ambiq MSPI driver, two intermediate chipset drivers for atxp032 and aps6404l and 2 samples for memc and flash.

Those will eventually be published here after MSPI is merged but maybe it would help the group if we were allowed to display them somewhere.

Reply:
> General question here- how do you envision supporting memory mapped access to MSPI peripherals? This is often the most efficient (and programmatically simple) method to configure XIP or interfacing with something like PSRAM.

I think @RichardSWheatley pretty much answered your question here. We do have a psram device driver planned for release in this PR.

Reply:
One other thing I would look at- we currently have a [FLEXSPI specific API](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/memc/memc_mcux_flexspi.h) within the tree that we use to support several PSRAM and flash devices. I don't think this API is a good candidate for a generic MSPI API, but it might be something to be aware of it you are interested in seeing how we are currently handling PSRAM

Reply:
> One other thing I would look at- we currently have a [FLEXSPI specific API](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/memc/memc_mcux_flexspi.h) within the tree that we use to support several PSRAM and flash devices. I don't think this API is a good candidate for a generic MSPI API, but it might be something to be aware of it you are interested in seeing how we are currently handling PSRAM

You should know that the FLEXSPI was the first thing I looked at when I moduled this MSPI API.

Reply:
> Those will eventually be published here after MSPI is merged but maybe it would help the group if we were allowed to display them somewhere.

This could indeed be useful to have a concrete usage example available to help converging on the API. This can be done as a side draft PR, which include current PR, and proposed as reference.

Reply:
I have one here if that helps you. https://github.com/AmbiqMicro/ambiqzephyr/pull/13